io-sim-classes/src/Control/Monad/Class/MonadSTM/Strict.hs

```diff
@@ -147,10 +145,22 @@ readTMVar StrictTMVar { tmvar } = Lazy.readTMVar tmvar
 tryReadTMVar :: MonadSTM m => StrictTMVar m a -> STM m (Maybe a)
 tryReadTMVar StrictTMVar { tmvar } = Lazy.tryReadTMVar tmvar
 
-swapTMVar :: MonadSTM m => StrictTMVar m a -> a -> STM m a
+swapTMVar :: (MonadSTM m, HasCallStack) => StrictTMVar m a -> a -> STM m a
 swapTMVar StrictTMVar { tmvar, invariant } !a =
-    assert (invariant a) $
+    checkInvariant (invariant a) $
     Lazy.swapTMVar tmvar a
 
 isEmptyTMVar :: MonadSTM m => StrictTMVar m a -> STM m Bool
 isEmptyTMVar StrictTMVar { tmvar } = Lazy.isEmptyTMVar tmvar
+
+{-------------------------------------------------------------------------------
+  Dealing with invariants
+-------------------------------------------------------------------------------}
+
+checkInvariant :: HasCallStack => Maybe String -> m x -> m x
```

Comment:
Would it be useful to reuse this function to check other invariants than this one? I mean instead of `assert`, which does not support a custom error message (or a call stack with a depth > 1).